src/Database/PostgreSQL/Simple/Notification.hs

```diff
@@ -120,9 +120,10 @@ getNotification conn = join $ withConnection conn fetch
                 -- the lock... but such a major bug is likely to exhibit
                 -- itself in an at least somewhat more dramatic fashion.)
                 Just fd  -> do
-                  (waitRead, _) <- threadWaitReadSTM fd
+                  (waitRead, unregister) <- threadWaitReadSTM fd
                   return $ do
-                    atomically waitRead `catch` (throwIO . setIOErrorLocation)
+                    mapException setIOErrorLocation
```

Comment:
why not

```haskell
atomically waitRead `catch` \e -> do
  unregister
  throwIO (setIOErrorLocation e)
```

or can `unregister` itself throw? Even then I wouldn't use `mapException` as it uses `unsafePerformIO`, and here it's not required.

Reply:
I think I tried this first and, strangely, it didn't resolve our issue but I'll try it again just to be sure. Agreed about `mapException`.

Reply:
I tested your suggestion but saw no change in the heap. I can still try to avoid `mapException` if you think it's that contentious (even though its documentation suggests it's safe).